### PR TITLE
[Validator] Pass validation groups on Collection validate.

### PR DESCRIPTION
**Pull request by Aliance**
| Q             | A
| ------------- | ---
| Branch?       | 2.7
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #23479
| License       | MIT

Complete description can be found in related issue.

### CODE REVIEW

**Comment by nicolas-grekas** [comment]
For some reason, Travis did not start. Please add tests also. The "BC break" flag means you should explain what's the precise behavior change, why you think we should still do it, explain why this cannot be done as a new feature on 3.4, with proper BC layer, etc.

**Comment by Aliance** [comment]
> For some reason, Travis did not start.

Because I created a pr into master, and then changed to 2.8. Can you restart them manually?

> Please add tests also.

Ok, I will try to find out where they should be.

> you should explain what's the precise behavior change

All description in linked issue.

**Comment by xabbuh** [comment]
Does this not happen on Symfony 2.7?

**Comment by Aliance** [comment]
@xabbuh I'am using LTS version 2.8, but I can check 2.7 a bit later.

**Comment by Aliance** [comment]
So, I check all tests, after my fix I have to fix `AbstractConstraintValidatorTest:: expectValidateValueAt`: now by default the `Default` validation group is passed to the mock method (see https://github.com/symfony/symfony/pull/23480/commits/b31795c72b68aaeca29c8fd5dcdf1ac232a92c91). Has not added any new tests yet, any ideas the best place for them?

Have no ideas why appveyor build fails, but seems there are some global problems, not because of my code.

While fixing the tests, found another place with the same bug: [AllValidator](https://github.com/symfony/symfony/blob/2.8/src/Symfony/Component/Validator/Constraints/AllValidator.php). Fix it too.

**Comment by Aliance** [comment]
> Does this not happen on Symfony 2.7?

I have checked it, the answer is **yes**, this bug is also actual in Symfony 2.7. I think it will reproduce on 2.5+ (see linked pr in issue)

**Comment by nicolas-grekas** [comment]
Can you rebase on 2.7 please?

**Comment by Aliance** [comment]
@nicolas-grekas done!

File: src/Symfony/Component/Validator/Tests/Constraints/AbstractConstraintValidatorTest.php [hunk 1]
```diff
@@ -30,6 +30,8 @@
  */
 abstract class AbstractConstraintValidatorTest extends TestCase
 {
+    const TEST_GROUP = 'MyGroup';
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Validator/Tests/Constraints/AbstractConstraintValidatorTest.php]: this should be reverted, it's not related to the fix

**Comment by nicolas-grekas** [comment]
@xabbuh WDYT?

**Comment by nicolas-grekas** [comment]
Just noticed that this looks a lot like #17696, maybe the discussion there would help move forward?

File: src/Symfony/Component/Validator/Tests/Constraints/AbstractConstraintValidatorTest.php (hunk 1 above)
Comment (Aliance) [inline src/Symfony/Component/Validator/Tests/Constraints/AbstractConstraintValidatorTest.php]: I have skipped the commit.

**Comment by Aliance** [comment]
I can't understand, why it pasts about 3 weeks since I've report a bug and make a PR that solves it, but it is still not merged. What should I do to speed up merging it?

**Comment by trilobit** [comment]
Hi! I have the same problem with validation groups. Let's merge this pull request.

**Comment by trilobit** [comment]
@nicolas-grekas @xabbuh what do you think?

**Comment by xabbuh** [comment]
The changes look valid to me. But can you add some test that would fail without them?

**Comment by Aliance** [comment]
@xabbuh I added some test (was taken from related pr), they fails in 2.7 without my fixes:

**Comment by Aliance** [comment]
@fabpot @nicolas-grekas @webmozart please look at my bugfix again and merge it please.

**Comment by fabpot** [comment]
You first need to rebase to get rid of  the merge commit.

**Comment by Aliance** [comment]
@fabpot I've already rebased, mb you see the cached commits?

**Comment by Aliance** [comment]
@fabpot @xabbuh @nicolas-grekas is there something wrong with this pr, why you don't reply to me? :)
Also, please, remove `BC break` label, this is not BC.

**Comment by javiereguiluz** [comment]
@Aliance I've removed the `BC Break`. About the lack of response, please be more patient. At the moment we're focused on [the Symfony 3.4 and Symfony 4.0 feature freeze](https://symfony.com/blog/symfony-3-4-and-4-0-are-about-to-enter-into-their-feature-freeze-period) which ends this weekend. Since your pull request is about a bug fix, we can work on it even during the feature freeze period. Thanks!

**Comment by Aliance** [comment]
May be someone can take a look at this bugfix now?

File: src/Symfony/Component/Validator/Constraints/AllValidator.php [hunk 2]
```diff
@@ -44,12 +44,12 @@ public function validate($value, Constraint $constraint)
             $validator = $context->getValidator()->inContext($context);
 
             foreach ($value as $key => $element) {
-                $validator->atPath('['.$key.']')->validate($element, $constraint->constraints);
+                $validator->atPath('['.$key.']')->validate($element, $constraint->constraints, $constraint->groups);
```
Comment (xabbuh) [inline src/Symfony/Component/Validator/Constraints/AllValidator.php]: I asked the same in #17696 too: Don't we have to use `$context->getGroup()` here to not accidentally widen the validation scope?

Comment (Aliance) [reply src/Symfony/Component/Validator/Constraints/AllValidator.php]: I think getters is better than properties, but there is `$constraint->constraints` not `$constraint->getConstraints()`. So I wrote the same way.
Should I change only groups, change both or no one?

Comment (xabbuh) [reply src/Symfony/Component/Validator/Constraints/AllValidator.php]: No, what I mean is fetching the groups from the context but not from the constraint? There may be more groups configured on the constraint than what groups are actually used to validate.

Comment (Aliance) [reply src/Symfony/Component/Validator/Constraints/AllValidator.php]: Nope, It's not the same. Please see my original issue: https://github.com/symfony/symfony/issues/23479
I have attached xdebug screenshots there.
Also I have post a link to my symfony standart fork with easy code, which is broken with current version of Symfony, but it works with my fixes. You can clone it and make sure that `$context->getGroup` will have only a `Default` string, while `$constraint->constraints` will have an array of needed groups: `["info", "Default", "Category]` for the `Category` entity as it was set in [validation.yml](https://github.com/Aliance/SymfonyTest/blob/86404b6899300aa6d74e7a7d1ee471c76cffa131/src/AppBundle/Resources/config/validation.yml#L17-L17)

Comment (xabbuh) [reply src/Symfony/Component/Validator/Constraints/AllValidator.php]: You are right that there is still a problem. But just passing the constraints here does not work either. For example, if you take your example project and modify the controller to only run validation in the `Default` and `modify` group, the `Length` constraint will still be evaluated which is only part of the `info` group.

**Comment by nicolas-grekas** [comment]
@Aliance how could we move forward here?

**Comment by Aliance** [comment]
@nicolas-grekas I have no idea, but 1 year bug in Symfony makes me sad 😢 
Does someone has any ideas?

**Comment by xabbuh** [comment]
see also #25888 which tries to fix the issue too (currently with the same approach)

**Comment by artursvonda** [comment]
Need any help with finishing this PR? Seems easy enough. Is changing `$constraint->groups` to fetching group from context only change required here?

**Comment by xabbuh** [comment]
Thank you all for your input and working on this. This is finally going to be fixed in #29499.

**Pull request closed without merging**